File: 5_WebApps/KhachLink/Services/ConflictResolutionService.cs
using VanAn.Shared.Domain;
using VanAn.KhachLink.Models;

namespace VanAn.KhachLink.Services
{
    public interface IConflictResolutionService
    {
        Task<ResolutionResult> ResolveOrderConflictAsync(OfflineOrderDto offlineOrder, Order? serverOrder);
        Task<ResolutionResult> ResolveCartConflictAsync(List<OfflineOrderItemDto> offlineItems, List<CartItem> serverItems);
        Task<bool> ValidateOrderAsync(OfflineOrderDto order);
        Task<bool> ValidateCartAsync(List<OfflineOrderItemDto> items);
        Task<ConflictReport> GenerateConflictReportAsync(OfflineOrderDto offlineOrder, Order? serverOrder);
    }

    public class ResolutionResult
    {
        public bool Success { get; set; }
        public ResolutionAction Action { get; set; }
        public string? Reason { get; set; }
        public OfflineOrderDto? MergedOrder { get; set; }
        public List<OfflineOrderItemDto>? MergedItems { get; set; }
        public List<string> Warnings { get; set; } = [];
    }

    public class ConflictReport
    {
        public string OrderId { get; set; } = string.Empty;
        public bool HasConflict { get; set; }
        public List<string> Conflicts { get; set; } = [];
        public List<string> Warnings { get; set; } = [];
        public ResolutionAction RecommendedAction { get; set; }
        public string Recommendation { get; set; } = string.Empty;
    }

    public class ConflictResolutionService(
        ILogger<ConflictResolutionService> logger,
        ISyncConflictResolver baseResolver) : IConflictResolutionService
    {
        private readonly ILogger<ConflictResolutionService> _logger = logger;
        private readonly ISyncConflictResolver _baseResolver = baseResolver;

        public async Task<ResolutionResult> ResolveOrderConflictAsync(OfflineOrderDto offlineOrder, Order? serverOrder)
        {
            ResolutionResult result = new();

            try
            {
                // Validate offline order first
                if (!await ValidateOrderAsync(offlineOrder))
                {
                    result.Success = false;
                    result.Action = ResolutionAction.Error;
                    result.Reason = "Offline order validation failed";
                    return result;
                }

                // Use base resolver for initial resolution
                ConflictResolution baseResolution = await _baseResolver.ResolveOrderConflictAsync(offlineOrder, serverOrder);

                if (!baseResolution.Success)
                {
                    result.Success = false;
                    result.Action = ResolutionAction.Error;
                    result.Reason = baseResolution.Reason;
                    return result;
                }

                // Apply enhanced logic based on action
                switch (baseResolution.Action)
                {
                    case ResolutionAction.UseOffline:
                        result = await ProcessUseOfflineAsync(offlineOrder, serverOrder);
                        break;

                    case ResolutionAction.UseServer:
                        result = await ProcessUseServerAsync(offlineOrder, serverOrder);
                        break;

                    case ResolutionAction.Merge:
                        result = await ProcessMergeAsync(offlineOrder, serverOrder);
                        break;
                    case ResolutionAction.Skip:
                        break;
                    case ResolutionAction.Error:
                        break;
                    default:
                        result.Success = false;
                        result.Action = ResolutionAction.Error;
                        result.Reason = "Unknown resolution action";
                        break;
                }

                _logger.LogInformation("Order conflict resolved: {OrderId}, Action: {Action}",
                    offlineOrder.Id, result.Action);
            }
            catch (Exception ex)
            {
                result.Success = false;
                result.Action = ResolutionAction.Error;
                result.Reason = ex.Message;
                _logger.LogError(ex, "Failed to resolve order conflict: {OrderId}", offlineOrder.Id);
            }

            return result;
        }

        public async Task<ResolutionResult> ResolveCartConflictAsync(List<OfflineOrderItemDto> offlineItems, List<CartItem> serverItems)
        {
            ResolutionResult result = new();

            try
            {
                // Validate cart items
                if (!await ValidateCartAsync(offlineItems))
                {
                    result.Success = false;
                    result.Action = ResolutionAction.Error;
                    result.Reason = "Offline cart validation failed";
                    return result;
                }

                // Use base resolver
                ConflictResolution baseResolution = await _baseResolver.ResolveCartConflictAsync(offlineItems, serverItems);

                if (!baseResolution.Success)
                {
                    result.Success = false;
                    result.Action = ResolutionAction.Error;
                    result.Reason = baseResolution.Reason;
                    return result;
                }

                // Process merged items
                if (baseResolution.MergedOrder != null)
                {
                    result.MergedItems = baseResolution.MergedOrder.Items;
                    result.Action = ResolutionAction.Merge;
                    result.Success = true;
                    result.Reason = baseResolution.Reason;

                    // Add warnings for potential issues
                    await CheckCartWarningsAsync(result.MergedItems, result);
                }

                _logger.LogInformation("Cart conflict resolved with {ItemCount} items",
                    result.MergedItems?.Count ?? 0);
            }
            catch (Exception ex)
            {
                result.Success = false;
                result.Action = ResolutionAction.Error;
                result.Reason = ex.Message;
                _logger.LogError(ex, "Failed to resolve cart conflict");
            }

            return result;
        }

        public async Task<bool> ValidateOrderAsync(OfflineOrderDto order)
        {
            try
            {
                // Basic validation
                if (string.IsNullOrEmpty(order.Id) || string.IsNullOrEmpty(order.CustomerId))
                {
                    return false;
                }

                if (order.Items == null || order.Items.Count == 0)
                {
                    return false;
                }

                // Validate items
                foreach (OfflineOrderItemDto item in order.Items)
                {
                    if (string.IsNullOrEmpty(item.ProductId) || item.Quantity <= 0 || item.UnitPrice <= 0)
                    {
                        return false;
                    }

                    if (Math.Abs(item.TotalPrice - (item.Quantity * item.UnitPrice)) > 0.01m)
                    {
                        return false;
                    }
                }

                // Validate total amount
                decimal calculatedTotal = order.Items.Sum(i => i.TotalPrice);
                return Math.Abs(calculatedTotal - order.TotalAmount) <= 0.01m;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to validate order: {OrderId}", order.Id);
                return false;
            }
        }

        public async Task<bool> ValidateCartAsync(List<OfflineOrderItemDto> items)
        {
            try
            {
                if (items == null || items.Count == 0)
                {
                    return true; // Empty cart is valid
                }

                foreach (OfflineOrderItemDto item in items)
                {
                    if (string.IsNullOrEmpty(item.ProductId) || item.Quantity <= 0 || item.UnitPrice <= 0)
                    {
                        return false;
                    }

                    if (Math.Abs(item.TotalPrice - (item.Quantity * item.UnitPrice)) > 0.01m)
                    {
                        return false;
                    }
                }

                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to validate cart items");
                return false;
            }
        }

        public async Task<ConflictReport> GenerateConflictReportAsync(OfflineOrderDto offlineOrder, Order? serverOrder)
        {
            ConflictReport report = new()
            {
                OrderId = offlineOrder.Id,
                HasConflict = false
            };

            try
            {
                if (serverOrder == null)
                {
                    report.RecommendedAction = ResolutionAction.UseOffline;
                    report.Recommendation = "No server order found - use offline order";
                    return report;
                }

                // Check for conflicts
                List<string> conflicts = [];
                List<string> warnings = [];

                // Timestamp conflict
                if (Math.Abs((offlineOrder.CreatedAt - serverOrder.CreatedAt).TotalMinutes) > 5)
                {
                    conflicts.Add($"Timestamp mismatch: Offline {offlineOrder.CreatedAt}, Server {serverOrder.CreatedAt}");
                    report.HasConflict = true;
                }

                // Total amount conflict
                if (Math.Abs(offlineOrder.TotalAmount - serverOrder.TotalAmount) > 0.01m)
                {
                    conflicts.Add($"Total amount mismatch: Offline {offlineOrder.TotalAmount}, Server {serverOrder.TotalAmount}");
                    report.HasConflict = true;
                }

                // Items conflict
                HashSet<string> offlineProductIds = offlineOrder.Items.Select(i => i.ProductId).ToHashSet();
                HashSet<string> serverProductIds = serverOrder.Items.Select(i => i.ProductId.ToString()).ToHashSet();

                if (!offlineProductIds.SetEquals(serverProductIds))
                {
                    conflicts.Add("Items differ between offline and server orders");
                    report.HasConflict = true;
                }

                // Status conflict
                if (offlineOrder.Status != serverOrder.Status.Value)
                {
                    warnings.Add($"Status differs: Offline {offlineOrder.Status}, Server {serverOrder.Status}");
                }

                report.Conflicts = conflicts;
                report.Warnings = warnings;

                // Recommend action
                if (report.HasConflict)
                {
                    if (offlineOrder.CreatedAt > serverOrder.CreatedAt)
                    {
                        report.RecommendedAction = ResolutionAction.UseOffline;
                        report.Recommendation = "Offline order is newer - prefer offline version";
                    }
                    else
                    {
                        report.RecommendedAction = ResolutionAction.Merge;
                        report.Recommendation = "Server order is newer - merge items to preserve data";
                    }
                }
                else
                {
                    report.RecommendedAction = ResolutionAction.UseServer;
                    report.Recommendation = "No significant conflicts - use server version";
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to generate conflict report: {OrderId}", offlineOrder.Id);
                report.HasConflict = true;
                report.RecommendedAction = ResolutionAction.Error;
                report.Recommendation = ex.Message;
            }

            return report;
        }

        private static async Task<ResolutionResult> ProcessUseOfflineAsync(OfflineOrderDto offlineOrder, Order? serverOrder)
        {
            ResolutionResult result = new()
            {
                Success = true,
                Action = ResolutionAction.UseOffline,
                Reason = "Using offline order",
                MergedOrder = offlineOrder
            };

            // Add warnings if server order exists
            if (serverOrder != null)
            {
                if (serverOrder.CreatedAt > offlineOrder.CreatedAt)
                {
                    result.Warnings.Add("Server order is newer but using offline version");
                }

                if (serverOrder.Status.Value != offlineOrder.Status)
                {
                    result.Warnings.Add($"Status differs: Server {serverOrder.Status}, Offline {offlineOrder.Status}");
                }
            }

            return result;
        }

        private static async Task<ResolutionResult> ProcessUseServerAsync(OfflineOrderDto offlineOrder, Order? serverOrder)
        {
            ResolutionResult result = new()
            {
                Success = true,
                Action = ResolutionAction.UseServer,
                Reason = "Using server order"
            };

            if (serverOrder != null)
            {
                result.MergedOrder = OfflineOrderDto.FromDomain(serverOrder);

                if (offlineOrder.CreatedAt > serverOrder.CreatedAt)
                {
                    result.Warnings.Add("Offline order is newer but using server version");
                }
            }
            else
            {
                result.Success = false;
                result.Action = ResolutionAction.Error;
                result.Reason = "Server order not found";
            }

            return result;
        }

        private async Task<ResolutionResult> ProcessMergeAsync(OfflineOrderDto offlineOrder, Order? serverOrder)
        {
            ResolutionResult result = new()
            {
                Success = true,
                Action = ResolutionAction.Merge,
                Reason = "Merging orders"
            };

            if (serverOrder != null)
            {
                // Create merged order
                OfflineOrderDto mergedOrder = new()
                {
                    Id = serverOrder.Id.ToString(),
                    CustomerId = serverOrder.CustomerId?.ToString() ?? string.Empty,
                    ShopId = serverOrder.TenantId.Value.ToString(),
                    Status = serverOrder.Status.Value,
                    CreatedAtTimestamp = ((DateTimeOffset)serverOrder.CreatedAt).ToUnixTimeMilliseconds(),
                    Items = []
                };

                // Merge items
                HashSet<string> processedProductIds = [];

                // Add server items first
                foreach (OrderItem serverItem in serverOrder.Items)
                {
                    mergedOrder.Items.Add(new OfflineOrderItemDto
                    {
                        ProductId = serverItem.ProductId.ToString(),
                        Quantity = serverItem.Quantity,
                        UnitPrice = serverItem.UnitPrice,
                        TotalPrice = serverItem.TotalPrice
                    });
                    _ = processedProductIds.Add(serverItem.ProductId.ToString());
                }

                // Add offline items not in server
                foreach (OfflineOrderItemDto offlineItem in offlineOrder.Items)
                {
                    if (!processedProductIds.Contains(offlineItem.ProductId))
                    {
                        mergedOrder.Items.Add(offlineItem);
                    }
                }

                // Recalculate total
                mergedOrder.TotalAmount = mergedOrder.Items.Sum(i => i.TotalPrice);

                result.MergedOrder = mergedOrder;

                // Add warnings
                if (Math.Abs((offlineOrder.CreatedAt - serverOrder.CreatedAt).TotalMinutes) > 5)
                {
                    result.Warnings.Add("Merging orders with significant timestamp difference");
                }
            }
            else
            {
                result.Success = false;
                result.Action = ResolutionAction.Error;
                result.Reason = "Server order not found for merge";
            }

            return result;
        }

        private static async Task CheckCartWarningsAsync(List<OfflineOrderItemDto> items, ResolutionResult result)
        {
            if (items.Count > 20)
            {
                result.Warnings.Add("Large number of items in cart may affect performance");
            }

            decimal totalValue = items.Sum(i => i.TotalPrice);
            if (totalValue > 1000000) // 1 million VND
            {
                result.Warnings.Add("High-value cart - consider confirming with customer");
            }

            List<OfflineOrderItemDto> highQuantityItems = items.Where(i => i.Quantity > 10).ToList();
            if (highQuantityItems.Count != 0)
            {
                result.Warnings.Add($"Items with high quantity: {string.Join(", ", highQuantityItems.Select(i => i.ProductId))}");
            }
        }
    }
}

File: 5_WebApps/KhachLink/Services/ConflictResolutionService.ts
export interface ConflictResolutionService {
    resolveOrderConflict(offlineOrder: Order, serverOrder?: Order): Promise<ConflictResolution>;
    resolveCartConflict(offlineItems: CartItem[], serverItems: CartItem[]): Promise<ConflictResolution>;
    validateOrder(order: Order): Promise<boolean>;
    validateCart(items: CartItem[]): Promise<boolean>;
    generateConflictReport(offlineOrder: Order, serverOrder?: Order): Promise<ConflictReport>;
}

export class ConflictResolutionServiceImpl implements ConflictResolutionService {
    async resolveOrderConflict(offlineOrder: Order, serverOrder?: Order): Promise<ConflictResolution> {
        const result: ConflictResolution = {
            success: false,
            action: 'error',
            reason: '',
            mergedOrder: null,
            mergedItems: null,
            warnings: []
        };
        
        try {
            // Validate offline order first
            if (!await this.validateOrder(offlineOrder)) {
                result.action = 'error';
                result.reason = 'Offline order validation failed';
                return result;
            }
            
            // Case 1: No server order exists - use offline
            if (!serverOrder) {
                result.success = true;
                result.action = 'useOffline';
                result.reason = 'No server order found';
                result.mergedOrder = offlineOrder;
                return result;
            }
            
            // Case 2: Server order is newer - use server
            if (new Date(serverOrder.createdAt) > new Date(offlineOrder.createdAt)) {
                result.success = true;
                result.action = 'useServer';
                result.reason = 'Server order is newer';
                result.mergedOrder = serverOrder;
                return result;
            }
            
            // Case 3: Same timestamp - merge items
            const timeDiff = Math.abs(new Date(offlineOrder.createdAt).getTime() - new Date(serverOrder.createdAt).getTime());
            if (timeDiff < 5000) { // 5 seconds
                const mergedOrder = await this.mergeOrders(offlineOrder, serverOrder);
                result.success = true;
                result.action = 'merge';
                result.reason = 'Orders created at same time - merging items';
                result.mergedOrder = mergedOrder;
                return result;
            }
            
            // Case 4: Offline order is newer - use offline
            result.success = true;
            result.action = 'useOffline';
            result.reason = 'Offline order is newer';
            result.mergedOrder = offlineOrder;
            return result;
        } catch (error) {
            result.success = false;
            result.action = 'error';
            result.reason = error.message;
            return result;
        }
    }
    
    async resolveCartConflict(offlineItems: CartItem[], serverItems: CartItem[]): Promise<ConflictResolution> {
        const result: ConflictResolution = {
            success: false,
            action: 'error',
            reason: '',
            mergedOrder: null,
            mergedItems: [],
            warnings: []
        };
        
        try {
            // Validate cart items
            if (!await this.validateCart(offlineItems)) {
                result.action = 'error';
                result.reason = 'Offline cart validation failed';
                return result;
            }
            
            // Merge cart items - combine unique items
            const mergedItems = this.mergeCartItems(offlineItems, serverItems);
            
            result.success = true;
            result.action = 'merge';
            result.reason = 'Cart items merged successfully';
            result.mergedItems = mergedItems;
            
            // Add warnings for potential issues
            this.checkCartWarnings(mergedItems, result);
            
            return result;
        } catch (error) {
            result.success = false;
            result.action = 'error';
            result.reason = error.message;
            return result;
        }
    }
    
    async validateOrder(order: Order): Promise<boolean> {
        try {
            // Basic validation
            if (!order.id || !order.customerId || !order.shopId) {
                return false;
            }
            
            if (!order.items || order.items.length === 0) {
                return false;
            }
            
            // Validate items
            for (const item of order.items) {
                if (!item.productId || item.quantity <= 0 || item.unitPrice <= 0) {
                    return false;
                }
                
                if (Math.abs(item.totalPrice - (item.quantity * item.unitPrice)) > 0.01) {
                    return false;
                }
            }
            
            // Validate total amount
            const calculatedTotal = order.items.reduce((sum, item) => sum + item.totalPrice, 0);
            if (Math.abs(calculatedTotal - order.totalAmount) > 0.01) {
                return false;
            }
            
            return true;
        } catch (error) {
            console.error('Order validation failed:', error);
            return false;
        }
    }
    
    async validateCart(items: CartItem[]): Promise<boolean> {
        try {
            if (!items || items.length === 0) {
                return true; // Empty cart is valid
            }
            
            for (const item of items) {
                if (!item.productId || item.quantity <= 0 || item.unitPrice <= 0) {
                    return false;
                }
                
                if (Math.abs(item.totalPrice - (item.quantity * item.unitPrice)) > 0.01) {
                    return false;
                }
            }
            
            return true;
        } catch (error) {
            console.error('Cart validation failed:', error);
            return false;
        }
    }
    
    async generateConflictReport(offlineOrder: Order, serverOrder?: Order): Promise<ConflictReport> {
        const report: ConflictReport = {
            orderId: offlineOrder.id,
            hasConflict: false,
            conflicts: [],
            warnings: [],
            recommendedAction: 'useOffline',
            recommendation: ''
        };
        
        try {
            if (!serverOrder) {
                report.recommendedAction = 'useOffline';
                report.recommendation = 'No server order found - use offline order';
                return report;
            }
            
            // Check for conflicts
            const timeDiff = Math.abs(new Date(offlineOrder.createdAt).getTime() - new Date(serverOrder.createdAt).getTime());
            
            if (timeDiff > 300000) { // 5 minutes
                report.conflicts.push(`Timestamp mismatch: Offline ${offlineOrder.createdAt}, Server ${serverOrder.createdAt}`);
                report.hasConflict = true;
            }
            
            if (Math.abs(offlineOrder.totalAmount - serverOrder.totalAmount) > 0.01) {
                report.conflicts.push(`Total amount mismatch: Offline ${offlineOrder.totalAmount}, Server ${serverOrder.totalAmount}`);
                report.hasConflict = true;
            }
            
            // Check items differences
            const offlineProductIds = offlineOrder.items.map(i => i.productId);
            const serverProductIds = serverOrder.items.map(i => i.productId);
            
            if (!this.arraysEqual(offlineProductIds, serverProductIds)) {
                report.conflicts.push('Items differ between offline and server orders');
                report.hasConflict = true;
            }
            
            // Status conflict
            if (offlineOrder.status !== serverOrder.status) {
                report.warnings.push(`Status differs: Offline ${offlineOrder.status}, Server ${serverOrder.status}`);
            }
            
            // Recommend action
            if (report.hasConflict) {
                if (new Date(offlineOrder.createdAt) > new Date(serverOrder.createdAt)) {
                    report.recommendedAction = 'useOffline';
                    report.recommendation = 'Offline order is newer - prefer offline version';
                } else {
                    report.recommendedAction = 'merge';
                    report.recommendation = 'Server order is newer - merge items to preserve data';
                }
            } else {
                report.recommendedAction = 'useServer';
                report.recommendation = 'No significant conflicts - use server version';
            }
        } catch (error) {
            report.hasConflict = true;
            report.recommendedAction = 'error';
            report.recommendation = error.message;
        }
        
        return report;
    }
    
    private async mergeOrders(offlineOrder: Order, serverOrder: Order): Promise<Order> {
        // Create merged order with server data as base
        const mergedOrder: Order = {
            id: serverOrder.id,
            customerId: serverOrder.customerId,
            shopId: serverOrder.shopId,
            totalAmount: serverOrder.totalAmount,
            status: serverOrder.status,
            createdAt: serverOrder.createdAt,
            items: []
        };
        
        // Merge items from both orders
        const processedProductIds = new Set<string>();
        
        // Add server items first
        for (const serverItem of serverOrder.items) {
            mergedOrder.items.push({
                productId: serverItem.productId,
                quantity: serverItem.quantity,
                unitPrice: serverItem.unitPrice,
                totalPrice: serverItem.totalPrice
            });
            processedProductIds.add(serverItem.productId);
        }
        
        // Add offline items that aren't in server
        for (const offlineItem of offlineOrder.items) {
            if (!processedProductIds.has(offlineItem.productId)) {
                mergedOrder.items.push({
                    productId: offlineItem.productId,
                    quantity: offlineItem.quantity,
                    unitPrice: offlineItem.unitPrice,
                    totalPrice: offlineItem.totalPrice
                });
            }
        }
        
        // Recalculate total
        mergedOrder.totalAmount = mergedOrder.items.reduce((sum, item) => sum + item.totalPrice, 0);
        
        return mergedOrder;
    }
    
    private mergeCartItems(offlineItems: CartItem[], serverItems: CartItem[]): CartItem[] {
        const mergedItems: CartItem[] = [];
        const processedProductIds = new Set<string>();
        
        // Add offline items first
        for (const offlineItem of offlineItems) {
            mergedItems.push(offlineItem);
            processedProductIds.add(offlineItem.productId);
        }
        
        // Add server items not in offline
        for (const serverItem of serverItems) {
            if (!processedProductIds.has(serverItem.productId)) {
                mergedItems.push({
                    productId: serverItem.productId,
                    quantity: serverItem.quantity,
                    unitPrice: serverItem.unitPrice,
                    totalPrice: serverItem.totalPrice
                });
            }
        }
        
        return mergedItems;
    }
    
    private checkCartWarnings(items: CartItem[], result: ConflictResolution): void {
        if (items.length > 20) {
            result.warnings.push('Large number of items in cart may affect performance');
        }
        
        const totalValue = items.reduce((sum, item) => sum + item.totalPrice, 0);
        if (totalValue > 1000000) { // 1 million VND
            result.warnings.push('High-value cart - consider confirming with customer');
        }
        
        const highQuantityItems = items.filter(item => item.quantity > 10);
        if (highQuantityItems.length > 0) {
            result.warnings.push(`Items with high quantity: ${highQuantityItems.map(i => i.productId).join(', ')}`);
        }
    }
    
    private arraysEqual(a: string[], b: string[]): boolean {
        if (a.length !== b.length) return false;
        const sortedA = [...a].sort();
        const sortedB = [...b].sort();
        return sortedA.every((val, index) => val === sortedB[index]);
    }
}

export interface ConflictResolution {
    success: boolean;
    action: 'useOffline' | 'useServer' | 'merge' | 'skip' | 'error';
    reason: string;
    mergedOrder?: Order;
    mergedItems?: CartItem[];
    warnings: string[];
}

export interface ConflictReport {
    orderId: string;
    hasConflict: boolean;
    conflicts: string[];
    warnings: string[];
    recommendedAction: 'useOffline' | 'useServer' | 'merge' | 'error';
    recommendation: string;
}

export interface Order {
    id: string;
    customerId: string;
    shopId: string;
    items: OrderItem[];
    totalAmount: number;
    status: string;
    createdAt: string;
    updatedAt?: string;
}

export interface OrderItem {
    productId: string;
    quantity: number;
    unitPrice: number;
    totalPrice: number;
}

export interface CartItem {
    productId: string;
    quantity: number;
    unitPrice: number;
    totalPrice: number;
}
